Add Linxura smart controller button quirk

simon3panda opened the pull request:
## Proposed change
<!--
  Explain your proposed change below.
-->
This PR adds a new quirk for the Linxura smart controller's Zigbee mode, enabling Linxura smart controller proper communication and control within the ZHA framework. Linxura smart controller's Zigbee mode support four Zigbee buttons. Each button support click, double click, long press actions.

## Additional information
<!--
  Please include any additional information that is important to this PR.
  For example, if this PR is a potentially breaking change, mention that here.
  If this PR requires other PRs to be merged in HA Core or other projects, mention that.
  Lastly, if this PR fixes a specific issue, please include "Fixes #xxxx".
-->
Test file is in tests/test_linxura.py


## Checklist
<!--
  Put an 'x' in all boxes that apply.
  Note: You do not need to tick all boxes before creating a PR.
-->

- [x] The changes are tested and work correctly
- [x] `pre-commit` checks pass / the code has been formatted using Black
- [x] Tests have been added to verify that the new code works

simon3panda committed:
first time to commit linxura integration

simon3panda committed:
Add new quirks that support Linxura smart controller.

Linxura smart controller's Zigbee mode support four buttons. Each button support click, double click, long press actions.

pre-commit-ci[bot] committed:
Apply pre-commit auto fixes

simon3panda committed:
Update test_linxura.py

modify spell issue

simon3panda committed:
Merge branch 'dev' of https://github.com/simon3panda/zha-device-handlers into dev

pre-commit-ci[bot] committed:
Apply pre-commit auto fixes

simon3panda committed:
fix pre-commit issue

simon3panda committed:
Merge branch 'dev' of https://github.com/simon3panda/zha-device-handlers into dev

simon3panda committed:
fix pre-commit issue

pre-commit-ci[bot] committed:
Apply pre-commit auto fixes

simon3panda committed:
fix pre-commit ruff-format issue

simon3panda committed:
Merge branch 'dev' of https://github.com/simon3panda/zha-device-handlers into dev

simon3panda committed:
codecov test issue

pre-commit-ci[bot] committed:
Apply pre-commit auto fixes

simon3panda committed:
fix pre-commit issue

simon3panda committed:
Merge branch 'dev' of https://github.com/simon3panda/zha-device-handlers into dev

simon3panda committed:
update for solving codecov cover issue

simon3panda committed:
solve check issue

pre-commit-ci[bot] committed:
Apply pre-commit auto fixes

simon3panda committed:
Update __init__.py

simon3panda committed:
Merge branch 'dev' of https://github.com/simon3panda/zha-device-handlers into dev

simon3panda committed:
Update __init__.py

pre-commit-ci[bot] committed:
Apply pre-commit auto fixes

simon3panda committed:
Update __init__.py

pre-commit-ci[bot] committed:
Apply pre-commit auto fixes

simon3panda committed:
Update __init__.py

pre-commit-ci[bot] committed:
Apply pre-commit auto fixes

simon3panda committed:
update __init__.py

simon3panda committed:
Update __init__.py

pre-commit-ci[bot] committed:
Apply pre-commit auto fixes

simon3panda committed:
Update __init__.py

simon3panda committed:
Update __init__.py

pre-commit-ci[bot] committed:
Apply pre-commit auto fixes

simon3panda committed:
Update __init__.py

codecov fix

simon3panda committed:
update for codecov cover

pre-commit-ci[bot] committed:
Apply pre-commit auto fixes

simon3panda committed:
Update test_linxura.py

simon3panda committed:
Merge branch 'dev' of https://github.com/simon3panda/zha-device-handlers into dev

pre-commit-ci[bot] committed:
Apply pre-commit auto fixes

simon3panda committed:
Update __init__.py

simon3panda committed:
Merge branch 'dev' of https://github.com/simon3panda/zha-device-handlers into dev

simon3panda committed:
Update test_linxura.py

simon3panda committed:
solve codecov issue

simon3panda committed:
Update test_linxura.py

pre-commit-ci[bot] committed:
Apply pre-commit auto fixes

simon3panda committed:
Update test_linxura.py

simon3panda committed:
Merge branch 'dev' of https://github.com/simon3panda/zha-device-handlers into dev

simon3panda committed:
Update __init__.py

simon3panda committed:
Update __init__.py

simon3panda committed:
Update __init__.py

simon3panda committed:
Update __init__.py

simon3panda committed:
Update __init__.py

simon3panda committed:
Update test_linxura.py

simon3panda committed:
Update test_linxura.py

simon3panda committed:
Update __init__.py

simon3panda committed:
Update __init__.py

TheJulianJES reviewed (commented):
Thanks for the PR. Please remove the commented out code.

simon3panda committed:
Remove the commented out code

simon3panda committed:
Update test_linxura.py

pre-commit-ci[bot] committed:
Apply pre-commit auto fixes

simon3panda commented:
> Thanks for the PR. Please remove the commented out code.
Thank you for reviewing the code, the comment code has been removed!

linxura1 commented:
@TheJulianJES Can you please merge this pull request? Thanks!

simon3panda commented:
> Thanks for the PR. Please remove the commented out code.

Hi Julian, I've uploaded the code that removed commented part, is there anything I need to do before merge?

TheJulianJES commented:
Have a bit of patience please. The PR needs to be reviewed properly. That'll be done when one of us has some free time.

simon3panda commented:
> Have a bit of patience please. The PR needs to be reviewed properly. That'll be done when one of us has some free time.

Sure！Thank you for your quick feedback.

simon3panda committed:
Modify 4 endpoint to 1 endpoint.

Linxura smart controller zigbee button, in order to meet persist feature, modify 4 endpoints to 1 endpoint. Still 4 buttons are shown on Home Assistant with press, double press, long press actions.

pre-commit-ci[bot] committed:
Apply pre-commit auto fixes

simon3panda committed:
Update __init__.py

simon3panda committed:
Merge branch 'dev' of https://github.com/simon3panda/zha-device-handlers into dev

simon3panda committed:
Create test_linxura.py

simon3panda committed:
Update test_linxura.py

simon3panda committed:
Delete test_linxura.py

simon3panda committed:
Create test_linxura.py

simon3panda committed:
Update test_linxura.py

simon3panda commented:
Linxura smart controller zigbee button, in order to meet persist feature, modify 4 endpoints to 1 endpoint. Still 4 buttons are shown on Home Assistant with press, double press, long press actions.
Please review, Thank you!

TheJulianJES commented on zhaquirks/linxura/__init__.py:
```diff
@@ -0,0 +1,57 @@
+    def _update_attribute(self, attrid, value):
+        self.info("Linxura update attribute - attrid: %d, value:%d", attrid, value)
+        super()._update_attribute(attrid, value)
+        if attrid == STATUS_REPORT:
```
Remove the constant `STATUS_REPORT` and replace it with getting the id of `zone_status` like this:
```suggestion
        if attrid == self.AttributeDefs.zone_status.id:
```
(or `IasZone.AttributeDefs.zone_status.id`)

simon3panda replied:
Thank you! Modified based on your kind suggestion.

TheJulianJES commented on zhaquirks/linxura/__init__.py:
```diff
@@ -0,0 +1,57 @@
+"""Linxura button device."""
+
+from zigpy.quirks import CustomCluster
+import zigpy.zcl.clusters.security
```
Just import IasZone here: `from zigpy.zcl.clusters.security import IasZone`.

```suggestion
import zigpy.zcl.clusters.security
from zigpy.zcl.clusters.security import IasZone
```

TheJulianJES commented on zhaquirks/linxura/__init__.py:
```diff
@@ -0,0 +1,57 @@
+}
+
+
+class LinxuraIASCluster(CustomCluster, zigpy.zcl.clusters.security.IasZone):
```
And if `IasZone` is imported, use this here:
```suggestion
class LinxuraIASCluster(CustomCluster, IasZone):
```

TheJulianJES replied:
This all seems to be relevant to only the one button quirk for now, so I'd move everything from this file into `button.py`. This follows what we do with other quirks.
Only if another quirk (file) also needs to access it, should it be moved here.
This file should just contain the docstring from the top then.

TheJulianJES commented on zhaquirks/linxura/__init__.py:
```diff
@@ -0,0 +1,57 @@
+    """Occupancy cluster."""
+
+    def _update_attribute(self, attrid, value):
+        self.info("Linxura update attribute - attrid: %d, value:%d", attrid, value)
```
Remove this logging line.

TheJulianJES commented on zhaquirks/linxura/__init__.py:
```diff
@@ -0,0 +1,57 @@
+
+
+class LinxuraIASCluster(CustomCluster, zigpy.zcl.clusters.security.IasZone):
+    """Occupancy cluster."""
```
```suggestion
    """IAS cluster used for Linxura button."""
```

TheJulianJES commented on tests/test_linxura.py:
```diff
@@ -0,0 +1,147 @@
+
+
+@pytest.mark.parametrize("quirk", (zhaquirks.linxura.button.LinxuraButton,))
+async def test_Linxura_button(zigpy_device_from_quirk, quirk):
```
Unless I'm missing something, this entire test could be compacted down to this:
```python
@pytest.mark.parametrize("value", (1, 3, 5, 7, 9, 11, 13, 15, 17, 19, 21, 23))
async def test_Linxura_button(zigpy_device_from_quirk, value):
    """Test Linxura button remotes."""

    device = zigpy_device_from_quirk(zhaquirks.linxura.button.LinxuraButton)

    # endpoint 1
    cluster = device.endpoints[1].ias_zone
    ias_zone_listener = ClusterListener(cluster)
    ias_zone_status_attr_id = IasZone.AttributeDefs.zone_status.id

    # button press
    cluster.update_attribute(ias_zone_status_attr_id, value)

    # ZHA calls update_attribute on the IasZone cluster when it receives a status_change_notification
    assert len(ias_zone_listener.attribute_updates) == 1
    assert ias_zone_listener.attribute_updates[0][0] == ias_zone_status_attr_id
    assert ias_zone_listener.attribute_updates[0][1] == value
```

TheJulianJES commented on tests/test_linxura.py:
```diff
@@ -0,0 +1,147 @@
+
+
+@pytest.mark.parametrize("quirk", (zhaquirks.linxura.button.LinxuraButton,))
+async def test_Linxura_button(zigpy_device_from_quirk, quirk):
```
Also, this test isn't testing the quirk. It just checks that the `update_attribute` call is passed and increases coverage by making sure all lines from the quirk are called, but we never verify the outputs..
We also can't and don't need to test ZHA here.

A proper test should verify that all cluster listeners with `ZHA_SEND_EVENT` are called as expected.
You can use something like this (and make use of `parametrize`):
```python
    listener = mock.MagicMock()
    cluster.add_listener(listener)

    assert cluster_listener.zha_send_event.call_count == 1
    assert cluster_listener.zha_send_event.call_args == mock.call(
        "button_1_press", {"button": "button_3", "press_type": "press", "command_id": 10}
    )
```

simon3panda replied:
Modified based on suggestion, however I'm a newbie in python. Didn't found a way to use parametrize properly to optimise and shorten the testing code. Please help me on this. Thanks a lot.

TheJulianJES replied:
All good! I've pushed a small change regarding this here: https://github.com/zigpy/zha-device-handlers/pull/3392/commits/b81c52d8418eab6310577c4ab5134c73dccad662

TheJulianJES commented on zhaquirks/linxura/__init__.py:
```diff
@@ -0,0 +1,57 @@
+                event_args = {
+                    BUTTON: button,
+                    PRESS_TYPE: press_type,
+                    COMMAND_ID: 10,  # to maintain backwards compatibility
```
To maintain backwards compatibility with what?
As far as I can see, these events were never fired before this PR.

If this was copy/pasted from another quirk and it's not needed, please remove it.

simon3panda replied:
Modified based on suggestion

TheJulianJES commented on zhaquirks/linxura/button.py:
```diff
@@ -0,0 +1,120 @@
+    }
+
+    device_automation_triggers = {
+        (SHORT_PRESS, BTN_1): {
```
I believe we should use `BUTTON_1` (and so on) from `zhaquirks.const` here.
We have translations for those in Home Assistant: [homeassistant/components/zha/strings.json#L240-L245](https://github.com/home-assistant/core/blob/d4071e7123d05acbda07864506541f879fb563cd/homeassistant/components/zha/strings.json#L240-L245)

simon3panda replied:
Modified based on suggestion.

TheJulianJES commented on zhaquirks/linxura/button.py:
```diff
@@ -0,0 +1,120 @@
+        }
+    }
+
+    device_automation_triggers = {
```
Since we're firing the events ourselves, we could use the following **if** we fire `SHORT_PRESS` (and so on) instead of `COMMAND_PRESS` in the `listener_event` for `ZHA_SEND_EVENT`.
```python
    device_automation_triggers = {
        (press_type, button): {
            COMMAND: f"{button}_{press_type}",
            CLUSTER_ID: IasZone.cluster_id,
        }
        for press_type in (SHORT_PRESS, DOUBLE_PRESS, LONG_PRESS)
        for button in (BUTTON_1, BUTTON_2, BUTTON_3, BUTTON_4)
    }
```

simon3panda replied:
Modified based on suggestion.

TheJulianJES commented on zhaquirks/linxura/__init__.py:
```diff
@@ -0,0 +1,57 @@
+
+
+PRESS_TYPES = {
+    1: COMMAND_PRESS,
+    2: COMMAND_DOUBLE,
+    3: COMMAND_HOLD,
```
Per the previous comment, we could fire these press types to have them match the trigger we provide to HA.
That would allow us to compact the `device_automation_triggers`.
```suggestion
    1: SHORT_PRESS,
    2: DOUBLE_PRESS,
    3: LONG_PRESS,
```

simon3panda committed:
Update __init__.py

Remove the constant STATUS_REPORT and replace it with getting the id of zone_status.

pre-commit-ci[bot] committed:
Apply pre-commit auto fixes

simon3panda committed:
Update zhaquirks/linxura/__init__.py

Co-authored-by: TheJulianJES <TheJulianJES@users.noreply.github.com>

simon3panda committed:
modify based on suggestion

pre-commit-ci[bot] committed:
Apply pre-commit auto fixes

simon3panda committed:
Update __init__.py

simon3panda committed:
Update __init__.py

simon3panda committed:
Update test_linxura.py

simon3panda committed:
Update test_linxura.py

pre-commit-ci[bot] committed:
Apply pre-commit auto fixes

simon3panda commented:
All suggested modifications has been made. Please review if you spare time. Many thanks!

TheJulianJES committed:
Clean up tests

TheJulianJES committed:
Simplify comparisons

TheJulianJES committed:
Use `elif`

TheJulianJES committed:
Move up `0 < value < 24` check

TheJulianJES committed:
Discard invalid values

TheJulianJES committed:
Clean up tests more

TheJulianJES commented on tests/test_linxura.py:
```diff
@@ -0,0 +1,213 @@
+    cluster.update_attribute(ias_zone_status_attr_id, 4)
+    assert (
+        len(ias_zone_listener.attribute_updates) == 1
+    )  # No update should occur for state >= 4
```
What is the purpose of this test? The comment states "No update should occur for state >= 4", but the IAS zone status attribute update always happens, regardless of "state".

TheJulianJES replied:
Update: I've reworked the tests and removed this for now.

TheJulianJES commented on zhaquirks/linxura/__init__.py:
```diff
@@ -0,0 +1,54 @@
+                event_args = {
+                    BUTTON: button,
+                    PRESS_TYPE: press_type,
+                    COMMAND_ID: 10,
```
I think we can remove the `COMMAND_ID` fully, right?

TheJulianJES replied:
Update: I've also removed this for now.

simon3panda replied:
I tested locally, it looks good! Thank you!

TheJulianJES committed:
Remove `COMMAND_ID`

TheJulianJES committed:
Improve comment on discarding values

TheJulianJES commented:
Thanks for making the changes!
I've made some improvements to the quirk + tests and made sure the code doesn't break when values 0, 6, 12, or 18 are passed in. Those values shouldn't generate events, right?
The `COMMAND_ID` is also removed from the event and the device triggers. I don't think we actually need it here.
I've also moved the custom IAS cluster to the button quirk file. This is just to be consistent with other quirks. We only need it in `button.py`, so let's keep it there.

Please have a look at the new code, let me know what you think, and try to see if everything works still.

TheJulianJES committed:
Move IAS cluster and constants to button quirk

TheJulianJES reviewed (approved):
Thanks for your contribution and for testing again!
I think everything looks good to go now.
